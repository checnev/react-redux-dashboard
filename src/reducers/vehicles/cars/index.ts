import {
  GET_CARS_START,
  GET_CARS_SUCCESS,
  GET_CARS_FAIL,
  CarsActionTypes,
} from 'actions/vehicles/cars';

export enum CarStatus {
  Free,
  Busy,
  Repair,
}

export interface Car {
  id: number;
  stateNumber: string;
  model: string;
  status: CarStatus;
  color: string;
  imageUrl: string;
  coordinats: number[];
}

export interface CarsState {
  cars: Car[];
  isFetching: boolean;
}

const initialState: CarsState  = {
  cars: [],
  isFetching: false,
};

const carsReducer = (
  state = initialState,
  action: CarsActionTypes
): CarsState => {
  switch (action.type) {
    case GET_CARS_START:
      return { ...state, isFetching: true };
    case GET_CARS_SUCCESS: 
      return {
        ...state,
        isFetching: false,
        cars: action.payload,
      };
    case GET_CARS_FAIL:
      return {
        ...state,
        isFetching: false,
        cars: [],
      };
    default:
      return { ...state };
  }
};

export default carsReducer;
