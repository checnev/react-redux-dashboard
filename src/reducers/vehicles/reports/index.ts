import {
  GET_CAR_REPORTS_START,
  GET_CAR_REPORTS_SUCCESS,
  GET_CAR_REPORTS_FAIL,
  CarReportsActionTypes
} from 'actions/vehicles/reports';

import type { Moment } from 'moment';
import type { Car } from 'reducers/vehicles/cars';

export interface CarReport {
  income: number;
  incidents: number;
  countOfLeases: number;
}

export interface CarDailyReport extends CarReport {
  date: Moment;
  carId: number;
}

export type CarReports = WeakMap<Car, ReadonlyArray<CarDailyReport>>;

export interface ReportsState {
  reports: CarReports | null;
  isFetching: boolean;
  requestId: string | null;
}

const initialState: ReportsState = {
  reports: null,
  isFetching: false,
  requestId: null,
};

const reportsReducer = (
  state = initialState,
  action: CarReportsActionTypes
): ReportsState => {

  switch (action.type) {
    case GET_CAR_REPORTS_START:
      return { 
        ...state,
        isFetching: true,
        requestId: action.requestId,
      };
    case GET_CAR_REPORTS_SUCCESS:
      if (state.requestId !== action.requestId) return state;
      return {
        ...state,
        isFetching: false,
        reports: action.payload,
        requestId: null,
      };
    case GET_CAR_REPORTS_FAIL:
      if (state.requestId !== action.requestId) return state;
      return {
        ...state,
        isFetching: false,
        reports: null,
        requestId: null,
      };
    default:
      return { ...state };
  }
};

export default reportsReducer;
