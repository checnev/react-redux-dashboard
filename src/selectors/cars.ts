import { createSelector } from 'reselect';
import type { RootState } from 'store';
import type { Car } from 'reducers/vehicles/cars';
import type { CarReport, CarReports } from 'reducers/vehicles/reports';

export interface CarWithReport extends Car {
  report?: CarReport;
}

export const carsSelector = (state: RootState) => state.vehicles.cars.cars;
export const reportsSelector = (state: RootState) => state.vehicles.reports.reports;

export const carWithReportsSelector = createSelector(
  carsSelector,
  reportsSelector,
  (cars: Car[], reports: CarReports | null) => {
    const carList: CarWithReport[] = [];

    for (let car of cars) {
      const carReports = reports?.get(car);
      const carData: CarWithReport = Object.assign({}, car);

      if (carReports) {
        carData.report = carReports.reduce((summary: CarReport, report) => {
          summary.income += report.income;
          summary.countOfLeases += report.countOfLeases;
          summary.incidents += report.incidents;

          return summary;
        }, {
          income: 0,
          countOfLeases: 0,
          incidents: 0,
        });
      }

      carList.push(carData);
    }

    return carList;
  }
);
