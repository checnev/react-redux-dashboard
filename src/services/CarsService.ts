import moment, { Moment } from 'moment';
import type { Car } from 'reducers/vehicles/cars';
import type { CarDailyReport, CarReports } from 'reducers/vehicles/reports';


class CarsService {
  private static cars: Car[] = [
    {
      id: 1,
      stateNumber: 'С 012 РС',
      model: 'Hyndai Solaris (2014)',
      status: 0,
      color: '#d76830',
      imageUrl: '/images/cars/hyundai-solaris-2014.jpg',
      coordinats: [56.835832, 60.615666],
    },
    {
      id: 2,
      stateNumber: 'М 123 ВС',
      model: 'Kia Rio (2014)',
      status: 1,
      color: '#808483',
      imageUrl: '/images/cars/kia-rio-2014.jpg',
      coordinats: [56.816553, 60.564730],
    },
    {
      id: 3,
      stateNumber: 'О 684 РТ',
      model: 'Ford Mondeo (2018)',
      status: 2,
      color: '#655c57',
      imageUrl: '/images/cars/ford-mondeo-2018.jpg',
      coordinats: [56.831751, 60.562233],
    },
  ];

  public static getCarById(id: string): Promise<Car | undefined> {

    return new Promise((resolve) => {
      setTimeout(() => {
        const car = CarsService.cars.find((car) => String(car.id) === id);
        resolve(car);
      }, 2500);
    });
  }
  public static getCars(): Promise<Car[]> {
    return new Promise((resolve) => {
      setTimeout(() => resolve(CarsService.cars), 2500);
    });
  }

  public static getCarReports(cars: Car[], from: Moment, to: Moment): Promise<CarReports> {
    const carsData: CarReports = new WeakMap();
    const days = Math.abs(to.diff(from, 'd'));
    
    cars.forEach((car) => {
      const data: CarDailyReport[] = Array.from(
        { length: days }, 
        (day, index) => ({
          carId: car.id,
          date: moment(from).add(index, 'd'),
          income: Math.floor(Math.abs(Math.cos(index + car.id) * 50000)),
          countOfLeases: 10 + index + car.id,
          incidents: Math.floor(index * 0.5) % car.id,
        })
      );
      carsData.set(car, data);
    });

    return new Promise((resolve) => {
      setTimeout(() => resolve(carsData), 2500);
    });
  }
}

export default CarsService;
