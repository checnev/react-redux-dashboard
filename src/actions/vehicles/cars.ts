import CarsService from 'services/CarsService';
import { push, CallHistoryMethodAction } from 'connected-react-router';

import type { RootState } from 'store';
import type { ThunkAction } from 'redux-thunk';
import type { Car } from 'reducers/vehicles/cars';

interface GetCarsStartAction {
  type: typeof GET_CARS_START;
}

interface GetCarsSuccessAction {
  type: typeof GET_CARS_SUCCESS;
  payload: Car[];
}

interface GetCarsFailAction {
  type: typeof GET_CARS_FAIL;
  payload: string;
}

export type CarsActionTypes = GetCarsStartAction 
  | GetCarsSuccessAction 
  | GetCarsFailAction
  | CallHistoryMethodAction;

export const GET_CARS_START = 'GET_CARS_START';
export const GET_CARS_SUCCESS = 'GET_CARS_SUCCESS';
export const GET_CARS_FAIL = 'GET_CARS_FAIL';

export const getCarById = (id: string): ThunkAction<void, RootState, unknown, CarsActionTypes> => (
  async (dispatch) => {
    dispatch({
      type: GET_CARS_START,
    });

    try {
      const car = await CarsService.getCarById(id);

      if (!car) throw new Error(`Car with id ${id} doesn't exist.`);

      dispatch({
        type: GET_CARS_SUCCESS,
        payload: [car],
      });
    } catch (err) {
      dispatch({
        type: GET_CARS_FAIL,
        payload: err.message,
      });

      dispatch(push('/404'));
    }
  }
);

export const getCars = (): ThunkAction<void, RootState, unknown, CarsActionTypes> => (
  async (dispatch) => {
    dispatch({
      type: GET_CARS_START,
    });

    try {
      const cars = await CarsService.getCars();
      dispatch({
        type: GET_CARS_SUCCESS,
        payload: cars,
      });
    } catch (err) {
      dispatch({
        type: GET_CARS_FAIL,
        payload: err.message,
      });
    }
  }
);
