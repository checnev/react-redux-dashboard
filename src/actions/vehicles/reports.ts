import CarsService from 'services/CarsService';
import { v4 as uuid4 } from 'uuid';

import type { Moment } from 'moment';
import type { RootState } from 'store';
import type { ThunkAction } from 'redux-thunk';
import type { Car } from 'reducers/vehicles/cars';
import type { CarReports } from 'reducers/vehicles/reports';

interface ActionWithRequestId {
  requestId: string | null;
}

interface GetCarReportsStartAction extends ActionWithRequestId {
  type: typeof GET_CAR_REPORTS_START;
}

interface GetCarReportsSuccessAction extends ActionWithRequestId {
  type: typeof GET_CAR_REPORTS_SUCCESS;
  payload: CarReports;
}

interface GetCarReportsFailAction extends ActionWithRequestId {
  type: typeof GET_CAR_REPORTS_FAIL;
  payload: string;
}

export type CarReportsActionTypes = GetCarReportsStartAction | GetCarReportsSuccessAction | GetCarReportsFailAction;

export const GET_CAR_REPORTS_START = 'GET_CAR_REPORTS_START';
export const GET_CAR_REPORTS_SUCCESS = 'GET_CAR_REPORTS_SUCCESS';
export const GET_CAR_REPORTS_FAIL = 'GET_CAR_REPORTS_FAIL';

export const getCarReports = (
  cars: Car[],
  from: Moment,
  to: Moment
): ThunkAction<void, RootState, unknown, CarReportsActionTypes> => (
  async (dispatch) => {
    const requestId = uuid4();
    dispatch({
      type: GET_CAR_REPORTS_START,
      requestId,
    });
    
    try {
      const data = await CarsService.getCarReports(cars, from, to);
      dispatch({
        type: GET_CAR_REPORTS_SUCCESS,
        payload: data,
        requestId,
      });
    } catch (err) {
      dispatch({
        type: GET_CAR_REPORTS_FAIL,
        payload: err.message,
        requestId,
      });
    }
  }
);
