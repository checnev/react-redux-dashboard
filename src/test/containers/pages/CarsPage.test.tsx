import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { waitFor, screen } from '@testing-library/react';
import pretty from 'pretty';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import state from 'test/data/initialState';
import CarsPageContainer from 'containers/pages/Cars';
import CarsService from 'services/CarsService';
import cars from 'test/data/cars';
import { createCarReports } from 'test/utils/reports';

const getCars = jest.fn().mockResolvedValue(cars);
const getCarReports = jest.fn().mockResolvedValue(
  createCarReports(cars, state.dashboard.from, state.dashboard.to)
);

CarsService.getCars = getCars;
CarsService.getCarReports = getCarReports;

const mockStore = configureMockStore([thunk]);
let store: any = null;

let container: any = null;
beforeEach(() => {
  container = document.createElement('div');
  document.body.append(container);

  store = mockStore(state);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe('CarsPageContainer', () => {

  const setUp = (component: React.ReactNode) => (
    <Provider store={store}>
      <Router>
        {component}
      </Router>
    </Provider>
  );

  it('CarsPageContainer render', async () => {
    await act(async () => {
      render(setUp(<CarsPageContainer />), container);
    });

    await waitFor(() => {
      // 210149 - income for kia rio
      expect(screen.getByText('210149')).toBeDefined();
    });

    expect(
      pretty(container.innerHTML)
    ).toMatchSnapshot();
  });

});
