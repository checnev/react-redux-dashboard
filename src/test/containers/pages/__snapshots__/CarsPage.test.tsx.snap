// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarsPageContainer CarsPageContainer render 1`] = `
"<div class=\\"MuiGrid-root makeStyles-root-1 MuiGrid-container MuiGrid-spacing-xs-3\\">
  <div class=\\"MuiGrid-root makeStyles-header-2 MuiGrid-item MuiGrid-grid-xs-12\\">
    <h1 class=\\"MuiTypography-root MuiTypography-h4\\">Автомобили</h1>
  </div>
  <div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12\\">
    <hr class=\\"MuiDivider-root\\">
  </div>
  <div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12\\">
    <div class=\\"MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded\\">
      <table class=\\"MuiTable-root makeStyles-root-3\\" aria-label=\\"car list table\\">
        <thead class=\\"MuiTableHead-root\\">
          <tr class=\\"MuiTableRow-root MuiTableRow-head\\">
            <th class=\\"MuiTableCell-root MuiTableCell-head\\" scope=\\"col\\">Гос. Номер</th>
            <th class=\\"MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter\\" scope=\\"col\\">Марка, Модель</th>
            <th class=\\"MuiTableCell-root MuiTableCell-head sortColumn MuiTableCell-alignCenter\\" scope=\\"col\\"><span class=\\"MuiButtonBase-root MuiTableSortLabel-root\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">Кол-во Аренд<svg class=\\"MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z\\"></path></svg></span></th>
            <th class=\\"MuiTableCell-root MuiTableCell-head sortColumn MuiTableCell-alignCenter\\" scope=\\"col\\"><span class=\\"MuiButtonBase-root MuiTableSortLabel-root\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">Доход<svg class=\\"MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z\\"></path></svg></span></th>
            <th class=\\"MuiTableCell-root MuiTableCell-head sortColumn MuiTableCell-alignCenter\\" scope=\\"col\\"><span class=\\"MuiButtonBase-root MuiTableSortLabel-root\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">Происшествий<svg class=\\"MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z\\"></path></svg></span></th>
            <th class=\\"MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter\\" scope=\\"col\\">Статус</th>
          </tr>
        </thead>
        <tbody class=\\"MuiTableBody-root\\">
          <tr class=\\"MuiTableRow-root\\">
            <th class=\\"MuiTableCell-root MuiTableCell-body\\" role=\\"cell\\" scope=\\"row\\">A 100 AA</th>
            <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary\\" href=\\"/cars/1\\">Hyundai solaris 2014</a></td>
            <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">98</td>
            <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">229889</td>
            <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">0</td>
            <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">Свободен</td>
          </tr>
          <tr class=\\"MuiTableRow-root\\">
            <th class=\\"MuiTableCell-root MuiTableCell-body\\" role=\\"cell\\" scope=\\"row\\">А 101 АА</th>
            <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary\\" href=\\"/cars/2\\">Kia Rio 2014</a></td>
            <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">105</td>
            <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">210149</td>
            <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">3</td>
            <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">Занят</td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>"
`;
